module/zfs/zfs_log.c

```diff
@@ -525,6 +526,125 @@ zfs_log_rename(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype, znode_t *sdzp,
+ *
+ * This means that we would need to force out an in-progress txg to disk and
+ * start a new one before writing any ZIL records. This would ensure that
+ * previous versions of ZFS which do not support these log txtypes would never
+ * encounter them during ZIL replay. Doing this would hurt performance.
+ *
+ * Alternatively, we could just activate the feature on a pool when these
+ * renameat2 flags get first used and leave it at that. This would render the
+ * pool read-only importable on implementations without the new feature flag,
+ * even when no new txtypes were present on-disk. This would result in
+ * effectively all Linux pools no longer being writeably on other systems,
+ * which isn't an acceptable state of affairs.
+ *
+ * Instead, we choose to rely on the fact that the ZIL is replayed in single-
+ * threaded mode before the dataset is mounted. This means we can represent the
+ * otherwise atomic operations as a series of plain good old txtypes known to
+ * all current OpenZFS implementations. To do that, we use the following
+ * functions (at least until more platforms implement renameat2).
+ *
+ *     zfs_log_rename_exchange
+ *     zfs_log_rename_whiteout
+ */
+
+void
+zfs_log_rename_exchange(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype,
+    znode_t *sdzp, const char *sname, znode_t *tdzp, const char *dname,
+    znode_t *szp)
+{
+#ifndef __FreeBSD__
+	zfs_dirlock_t *tmpdl;
+#endif
+	znode_t *tmpzp = NULL;
+	char *tmpname;
+	int retries = 0;
+	int pos = 0;
+	int err;
+
+	if (zil_replaying(zilog, tx))
+		return;
+
+	tmpname = kmem_alloc(MAXPATHLEN, KM_SLEEP);
+
+	/*
+	 * To represent atomic rename with old non-atomic operations, we need a
+	 * temporary new name; so we try picking a name until we succeed, then
+	 * we get a dirent lock for that temp name until the final itx gets
+	 * queued. The 16-bytes-of-entropy random suffix means an attacker
+	 * would need to create 256^16 files (~10^38) files in order to cause
+	 * us to fail this.
+	 */
+retry:
+	retries++;
+	pos = snprintf(tmpname, MAXPATHLEN, "%s.zfs_renameat2_emul_", dname);
+
+	for (int i = 0; i < 16; i++) {
+		int r = 0xFF;
+		random_get_pseudo_bytes((void *)&r, 1);
+		pos += snprintf(tmpname+pos, MAXPATHLEN-pos, "%02x", r);
+	}
+
+#ifdef __FreeBSD__
+	/*
+	 * NOTE: This code is never called on FreeBSD because they don't have
+	 *       RENAME_EXCHANGE, so it is possible this code is missing
+	 *       correct locking (though as far as I can tell, that is not the
+	 *       case).
+	 */
+	err = zfs_dirent_lookup(tdzp, tmpname, &tmpzp, ZNEW);
+#else
+	err = zfs_dirent_lock(&tmpdl, tdzp, tmpname, &tmpzp, ZNEW, NULL, NULL);
+#endif
+
+	VERIFY3U(retries, <, 10);
+	if (err)
+		goto retry;
+
+	/* dst -> tmp */
+	zfs_log_rename(zilog, tx, txtype, tdzp, dname, tdzp, tmpname, szp);
+	/* src -> dst */
+	zfs_log_rename(zilog, tx, txtype, sdzp, sname, tdzp, dname, szp);
+	/* tmp -> src */
+	zfs_log_rename(zilog, tx, txtype, tdzp, tmpname, sdzp, sname, szp);
+
```

Comment:
There is no guarantee that the ITXs created by the functions above will land in the same LWB.
And even if they did, it's not guaranteed that an LWB will be replayed atomically.
Why?
* Replay up to the Nth log record of the LWB
* The txg with those replay effects syncs
* Crash
* Bitflip in LWB
* Reboot & pool import
* Replay will interpret corrupted LWB as end-of-chain ( other problems resulting from this design: https://github.com/openzfs/zfs/issues/11364 and https://github.com/openzfs/zfs/issues/11363 )
* We have a partially-replayed `zfs_log_rename_exchange` in the dataset.